packages/react-server/src/ReactFlightServer.js

```diff
@@ -3107,12 +3104,28 @@ function flushCompletedChunks(
   }
 }
 
+function completeWorkEpoch(request: Request) {
+  request.epoch++;
+  const destination = request.destination;
+  if (destination) {
+    flushCompletedChunks(request, destination);
+  }
+}
+
+function startPerformingWork(request: Request): void {
+  request.epoch++;
+  scheduleWork(() => performWork(request));
+  scheduleWork(() => {
+    completeWorkEpoch(request);
+  });
+}
+
 export function startWork(request: Request): void {
   request.flushScheduled = request.destination !== null;
   if (supportsRequestStorage) {
-    scheduleWork(() => requestStorage.run(request, performWork, request));
+    requestStorage.run(request, startPerformingWork, request);
```

Comment:
I don’t quite trust this thing here. It assume a lot about the scheduling correctly following the local storage implementation. Which 1) it’s implemented as AsyncLocalStorage at all which might not be the case 2) that the scheduling is implemented using something that follows the storage 3) the scheduling isn’t batched across requests.